man/rc-depend.8
.TH "BASELAYOUT" "8" "Oct 2006" "baselayout" "baselayout"
.SH NAME
rc-depend \ - resolve init script dependencies
.SH SYNOPSIS
\fBrc-depend\fR \fI-ineed\fR \fI-iuse\fR \fIservice\fR ...
.br
\fBrc-depend\fR \fI--notrace\fR \fI-iprovide\fR \fIservice\fR ...
.br
\fBrc-depend\fR \fI-needsme\fR \fIservice\fR ...
.SH DESCRIPTION
Gentoos init system uses service dependencies to depend on other services.
Rather than just starting in a set order, we start and stop in the order
defined by the services themselves.
For example, most services require local disks to be mounted and as such can
depend on the localmount service. Others depend on and net and dns and will
only start when those dependencies have been satisfied.

One issue of note is that a service can provide another service, which is more
generic. A good example of this is that net.lo and any service linked to it 
provide "net", which a few services depend on. You can of course have a few
network interfaces: modern laptops have 3 being loopback, wired and wireless.
What makes this more interesting is that it could be setup so that both wired
and wireless are optional. So we work out provided services like so :-

1) Always use any services in the runlevel.
.br
2) If no services are defined in the runlevel then use any running services
that satisfy the provide.
.br
3) Append any services in the boot runlevel.

\fBrc-order\fR is primarily used internally by Gentoo and is not meant as an
end-user or admin tool. This man page is purely to describe its function.
.SH OPTIONS
.TP
\fB--alwaysvalid\fR
Every service is valid. This is a needed option if the deptree is generated
from Gentoo network modules.
.TP
\fB--deptree \fIdeptree\fR
Use this \fIdeptree\fR instead of the default one,
\fI/lib/rcscripts/init.d/deptree\fR.
.TP
\fB--notrace\fR
Just show the dependencies for the specified services without working out
anything extra.
.TP
\fB--strict\fR
For provided services, depend on all of them in the runlevel instead of just
ones that are started.
.TP
\fB-dependency_type\fR
Work with the specified dependency type, such as \fIineed\fR, \fIiafter\fR,
\fIneedsme\fR.
If none are supplied we default to \fIineed\fR and \fIiuse\fR.
.SH NOTES
When needsme depends on a provided service, like net, we don't do any
mapping to an actual service unless it's the last one up. So if net.lo and
net.eth0 are started then neither are returned. If net.eth0 then stops then
every service that needs net then has net.lo in its needsme list so that
we net.lo stops it brings down all services that depend on net.
.SH BUGS
\fBrc-depend\fR relies on \fI/lib/rcscripts/init.d/deptree\fR being up-to-date.
Every time a service is run in /etc/init.d or when \fBrc\fR runs then the
\fIdeptree\fR is updated by \fBdepscan.sh\fR if any config files are newer
than the \fIdeptree\fR.
So in theory we should never have a problem with this.
.br
Provided services are calculated at runtime. The current downside of this
approach means that if you do "after net; before net.lo" and net.lo provides
net then you can get into an sticky loop where services hang.
.SH "REPORTING BUGS"
Please report bugs via http://bugs.gentoo.org/
.SH "SEE ALSO"
.BR rc-update (8),
.BR rc-status (8)
.SH AUTHORS
Roy Marples <uberlord@gentoo.org>


ChangeLog
# ChangeLog for Gentoo System Intialization ("rc") scripts
# Copyright 1999-2007 Gentoo Foundation; Distributed under the GPLv2

  18 Jan 2007; Mike Frysinger <vapier@gentoo.org>:

    Add /bin/fish to /etc/shells #162673 by Donnie Berkholz.

* baselayout-1.13.0_alpha12 (17 Jan 2006)

  16 Jan 2007; Roy Marples <uberlord@gentoo.org>:

    rc now stops services that are "starting" when in the stopping phase.

    Stop start-stop-daemon --quiet --test from spamming stdout.

    Stop hotplug events from spamming log files.

  15 Jan 2007; Roy Marples <uberlord@gentoo.org>:

    Fix ifplugd stopping.

    rc-status no longer errors when in single user on FreeBSD.

    rc-depend has a much saner calculates which services provide what
    in a much saner way. Hopefully this now works for all situations of
    how you want the net depend to be fulfilled!

  14 Jan 2007; Roy Marples <uberlord@gentoo.org>:

    Change from using symlinks to real files so we can test mtimes
    in svc_in_control, #154670.

    Fix ip6 routes when iproute2 isn't installed, #161959.

    Allow the setting of a global metric, #161822.

  12 Jan 2007; Roy Marples <uberlord@gentoo.org>:

    Revive old default behaviour of network scripts, #161510.
    This is toggled throught the use of RC_STRICT_DEPEND.

    Fix bootmisc, #161328.

  10 Jan 2007; Roy Marples <uberlord@gentoo.org>:

    Fix some restarting issues when --nodeps is used.
    Fix stopping for daemons that take some time.

  10 Jan 2007; Mike Frysinger <vapier@gentoo.org>:

    Rename --update option in depscan.sh to --force.

* baselayout-1.13.0_alpha11 (10 Jan 2006)

  10 Jan 2007; Roy Marples <uberlord@gentoo.org>:

    New Makefile directive FILES_APPEND so we can have a base config
    which OS specific configs can append to.

  09 Jan 2007; Mike Frysinger <vapier@gentoo.org>:

    If the adjtime file is not specified, then don't bother trying to use its
    facilities so that we can run clock as early as possible in the boot
    process.

  09 Jan 2007; Roy Marples <uberlord@gentoo.org>:

    Removed lots of rc-daemon.sh cruft. We now rely pretty much entirely on
    start-stop-daemon. This has the benefit of being smaller, no more
    dependency on pidof and more consistent and effcient.

    depscan.sh will correctly adjust future mtimes in /etc/conf.d and
    /etc/init.d, #157432 hopefully for the last time.

    Split clock configs into BSD and Linux - we share the same init file though.

    You can now specify relay and suffix for ip6to4 interfaces, thanks
    to Asfand Yar Qazi.

    Add conditionalrestart action, which is the same as restart but the service
    is only restarted if it is already started. This means that anything that
    whats this only has to use one command instead of two, making them more
    efficient.

  08 Jan 2007; Roy Marples <uberlord@gentoo.org>:

    Create a test file, and if the status is newer than that bail out
    after running the service, #154670.

    bootmisc now just checks to see if the pid exists for .pid files in
    /var/run as we cannot rely on the pidfile name matching the process
    name.

  05 Jan 2007; Roy Marples <uberlord@gentoo.org>:

    We now default to using WEP open unless specified.

    prism54 drivers always report a carrier, which is wrong. Luckily
    they don't report an essid when not assoicated, so we check that too.

    If we don't find an access point in the first scan, sleep for a second
    and try again. Do this another time by default for a total of 3 scans.
    This is in addition to the scans_eth0 setting. BSD has decent drivers,
    so no need for this there.

  05 Jan 2007; Mike Frysinger <vapier@gentoo.org>:

    Make sure symlinks to directories in /etc/init.d do not screw us up
    when we run depscan.sh #159999 by Majin Sniper.

  04 Jan 2007; Roy Marples <uberlord@gentoo.org>:

    scans_eth0 now states how many times we will run a scan before using
    the results, default 1. Fixes #159723.

    Add a new carrier_timeout_eth0 setting just for Intel e1000 users on
    FreeBSD where the carrier takes far too long to come up at boot.
    This is very likely a hardware problem as Linux users have reported
    similar issues with netplug/ifplugd which does the same thing here.

  03 Jan 2007; Roy Marples <uberlord@gentoo.org>:

    FreeBSD users can now specify a kernel dump device in conf.d/localmount
    Fixes #159643, thanks to Diego Pettenò.

    Save last used consolefont in /lib/rcscripts/console and apply
    in init.sh, #156604.

    Update ESSID/MAC wireless documentation, #159807 thanks to Michael Conrad.

  03 Jan 2007; Mike Frysinger <vapier@gentoo.org>:

    Add some more useful flags to rcscripts and document them (--debug/--help).

  30 Dec 2006; Mike Frysinger <vapier@gentoo.org>:

    Punt the old MAKEDEV script and move to MAKEDEV package from Fedora.

  29 Dec 2006; Roy Marples <uberlord@gentoo.org>:

    Use correct interface in example, #159396.

  19 Dec 2006; Roy Marples <uberlord@gentoo.org>:

    start-stop-daemon now works with pidfiles that put extra data after the pid
    like postgresql.

    Only set remotename if we have a username, #104869 thanks to Alin Nastac.

    Use the mtime of /etc/{conf,init}.d directories too so depscan.sh works
    for newly added files. Partial fix for #158548.

    Only launch sulogin if we are actually in single user mode, #158615.

* baselayout-1.13.0_alpha10 (12 Dec 2006)

  18 Dec 2006; Roy Marples <uberlord@gentoo.org>:

    Reverted patch for #104869 as it's silly. Fixes #158174.


src/rc-depend.c
/*
   rc-depend
   rc service dependency and ordering
   Copyright 2006-2007 Gentoo Foundation
   Written by Roy Marples <uberlord@gentoo.org>

   For optimum operation, we need to get several variables.
   The easiest way is via the environment, so these should be set before
   running rc-order :-
   SOFTLEVEL
   BOOTLEVEL

   We can also handle other dependency files and types, like Gentoos
   net modules. These are used with the --deptree and --awlaysvalid flags.
   */

#define SVCDIR 		"/lib/rcscripts/init.d"
#define DEPTREE 	SVCDIR "/deptree"
#define RUNLEVELDIR 	"/etc/runlevels"
#define INITDIR		"/etc/init.d"

#define BOOTLEVEL	"boot"
#define SOFTLEVEL	"default"
#define RCNETCHECK	"no"

#define MAXTYPES 	20 /* We currently only go upto 10 (providedby) */
#define LINEBUFFER	2048

#include <sys/types.h>
#include <sys/stat.h>

#include <err.h>
#include <errno.h>
#include <limits.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

char *bootlevel;
char *softlevel;
char *svcname;
bool always_valid = false;
bool strict = false;

struct linkedlist
{
  char *item;
  struct linkedlist *next;
} linkedlist;

struct deptype
{
  char *type;
  char *services;
  struct deptype *next;
} deptype;

struct depinfo
{
  char *service;
  struct deptype *deps;
  struct depinfo *next;
} depinfo;

void *xmalloc (size_t size)
{
  register void *value = malloc (size);

  if (value)
    return value;

  errx (EXIT_FAILURE, "memory exhausted");
}

struct linkedlist *add_linkedlist (struct linkedlist *linkedlist,
				   const char *item)
{
  if (! linkedlist || ! item)
    return NULL;

  struct linkedlist *p = linkedlist;
  if (! p->item)
    {
      p->item = strdup (item);
      return p;
    }

  while (p->next)
    p = p->next;

  p->next = xmalloc (sizeof (struct linkedlist));
  p = p->next;
  memset (p, 0, sizeof (struct linkedlist));
  p->item = strdup (item);

  return p;
}

void free_linkedlist (struct linkedlist *linkedlist)
{
  struct linkedlist *p = linkedlist;
  struct linkedlist *n = NULL;

  while (p)
    {
      n = p->next;
      if (p->item)
	free (p->item);
      free (p);
      p = n;
    }
}

char *get_shell_value (char *string)
{
  if (! string)
    return NULL;

  char *p = string;
  if (*p == '"')
    p++;
  char *e = p + strlen (p) - 1;
  if (*e == '\n')
    *e-- = 0;
  if (*e == '"')
    *e-- = 0;

  if (*p != 0)
    return p;

  return NULL;
}

bool is_runlevel_start ()
{
  struct stat sb;

  if (stat (SVCDIR "/softscripts.old", &sb) == 0)
    return true;
  return false;
}

bool is_runlevel_stop ()
{
  struct stat sb;

  if (stat (SVCDIR "/softscripts.new", &sb) == 0)
    return true;
  return false;
}

bool exists_dir_file (const char *dir, const char *file)
{
  int l = strlen (dir) + strlen (file) + 2;
  char *f = xmalloc (l);
  snprintf (f, l, "%s/%s", dir, file);

  bool result = false;
  struct stat sb;
  if (stat (f, &sb) == 0)
    result = true; 
  free (f);
  return (result);
}

bool exists_dir_dir_file (const char *dir1, const char *dir2, const char *file)
{
  int l = strlen (dir1) + strlen (dir2) + 2;
  char *d = xmalloc (l);
  snprintf (d, l, "%s/%s", dir1, dir2);

  bool result = exists_dir_file (d, file);
  free (d);
  return (result);
}

bool in_runlevel (const char *level, const char *service)
{
  return exists_dir_dir_file (RUNLEVELDIR, level, service);
}

bool service_coldplugged (const char *service)
{
  return exists_dir_dir_file (SVCDIR, "coldplugged", service);
}

bool service_starting (const char *service)
{
  return exists_dir_dir_file (SVCDIR, "starting", service);
}

bool service_inactive (const char *service)
{
  return exists_dir_dir_file (SVCDIR, "inactive", service);
}

bool service_started (const char *service)
{
  return exists_dir_dir_file (SVCDIR, "started", service);
}

bool service_stopping (const char *service)
{
  return exists_dir_dir_file (SVCDIR, "stopping", service);
}

bool service_stopped (const char *service)
{
  if (service_starting (service)
      || service_started (service)
      || service_inactive (service)
      || service_stopping (service))
    return false;

  return true;
}


void free_deptree (struct depinfo *deptree)
{
  struct depinfo *di = deptree;
  while (di)
    {
      free (di->service);
      struct depinfo *dip = di->next;
      struct deptype *dt = di->deps;
      while (dt)
	{
	  free (dt->type);  
	  free (dt->services);
	  struct deptype *dtp = dt->next;
	  free (dt);
	  dt = dtp;
	}
      free (di);
      di = dip;
    }
}

/* Load our deptree
   Although the deptree file is pure sh, it is in a fixed format created
   by gendeptree awk script. We depend on this, if you pardon the pun ;)
   */
struct depinfo *load_deptree (char *file)
{
  FILE *fp;
  if (! (fp = fopen (file, "r")))
    err (EXIT_FAILURE, "Failed to open deptree `%s'", file);

  char *types[MAXTYPES];
  memset (types, 0, MAXTYPES);

  struct depinfo *deptree = xmalloc (sizeof (struct depinfo));
  memset (deptree, 0, sizeof (struct depinfo));
  struct depinfo *depinfo = NULL;
  struct deptype *deptype = NULL;
  char buffer [LINEBUFFER];
  int rc_type_len = strlen ("declare -r rc_type_");
  int rc_depend_tree_len = strlen ("RC_DEPEND_TREE[");
  int max_type = 0;
  char *p, *e, *f;

  while (fgets (buffer, LINEBUFFER, fp))
    {
      /* Grab our types first */
      if (strncmp (buffer, "declare -r rc_type_", rc_type_len) == 0)
	{
	  p = buffer + rc_type_len;
	  if (! (e = strchr(p, '=')))
	    continue;

	  /* Blank out the = sign so we can just copy the text later */
	  *e = 0;
	  e++;

	  errno = 0;
	  long t = strtol (e, &f, 10);
	  if ((errno == ERANGE && (t == LONG_MAX || t == LONG_MIN))
	      || (errno != 0 && t == 0))
	    continue;

	  types[t] = strdup (p);
	  if (t > max_type)
	    max_type = t;

	  continue;
	}

      if (strncmp (buffer, "RC_DEPEND_TREE[", rc_depend_tree_len))
	continue;

      p = buffer + rc_depend_tree_len;
      e = NULL;

      errno = 0;
      long idx = strtol (p, &e, 10);
      if ((errno == ERANGE && (idx == LONG_MAX || idx == LONG_MIN))
	  || (errno != 0 && idx == 0))
	{
	  warnx ("load_deptree: `%s' is not an index", p);
	  continue;
	}

      if (idx == 0)
	continue;

      /* If we don't have a + then we're a new service
	 OK, this is a hack, but it works :) */
      if (*e == ']')
	{
	  e += 2; // ]=
	  if (! depinfo)
	    depinfo = deptree;
	  else
	    {
	      depinfo->next = xmalloc (sizeof (struct depinfo));
	      depinfo = depinfo->next;
	      memset (depinfo, 0, sizeof (struct depinfo));
	    }
	  deptype = NULL;
	  depinfo->service = strdup (get_shell_value (e));
	  continue;
	}

      /* Sanity */
      if (*e != '+')
	{
	  warnx ("load_deptree: expecting `+', got `%s'", e);
	  continue;
	}

      /* Now we need to work out our service value */
      p = e + 1;
      errno = 0;
      long val = strtol (p, &e, 10);
      if ((errno == ERANGE && (val == LONG_MAX || val == LONG_MIN))
	  || (errno != 0 && val == 0))
	{
	  warnx ("load_deptree: `%s' is not an service type", p);
	  continue;
	}

      if (! types[val])
	{
	  warnx ("load_deptree: we don't value a type for index `%li'", val);
	  continue;
	}

      if (*e != ']')
	{
	  warnx ("load_deptree: expecting `]', got `%s'", e);
	  continue;
	}
      e++;
      if (*e != '=')
	{
	  warnx ("load_deptree: expecting `=', got `%s'", e);
	  continue;
	}
      e++;

      /* If we don't have a value then don't bother to add the dep */
      char *x = get_shell_value (e++);
      if (! x)
	continue;

      if (deptype)
	{
	  deptype->next = xmalloc (sizeof (struct deptype));
	  deptype = deptype->next;
	}
      else
	{
	  depinfo->deps = xmalloc (sizeof (struct deptype));
	  deptype = depinfo->deps;
	}
      memset (deptype, 0, sizeof (struct deptype));

      deptype->type = strdup (types[val]);
      deptype->services = strdup (x);
    }

  fclose (fp);

  if (! depinfo)
    {
      free (deptree);
      deptree = NULL;
    }

  int i;
  for (i = 0; i <= max_type; i++)
    if (types[i])
      free (types[i]);

  return deptree;
}

struct depinfo *get_depinfo (struct depinfo *deptree, const char *service)
{
  if (! deptree || ! service)
    return NULL;

  struct depinfo *di;
  for (di = deptree; di; di = di->next)
    if (strcmp (di->service, service) == 0)
      return di;

  return NULL;
}

struct deptype *get_deptype (struct depinfo *depinfo, const char *type)
{
  if (! depinfo || !type)
    return NULL;

  struct deptype *dt;
  for (dt = depinfo->deps; dt; dt = dt->next)
    if (strcmp (dt->type, type) == 0)
      return dt;

  return NULL;
}

bool valid_service (const char *service)
{
  return (always_valid
	  || exists_dir_dir_file (RUNLEVELDIR, bootlevel, service)
	  || exists_dir_dir_file (RUNLEVELDIR, softlevel, service)
	  || service_coldplugged (service) || service_started (service));
}

bool get_provided1 (struct linkedlist *providers,
		    struct deptype *deptype,
		    char *level, bool coldplugged,
		    bool started, bool inactive)
{
  char *service;
  char *p;
  char *op;
  struct linkedlist *lp = providers;
  bool retval = false;

  op = p = strdup (deptype->services);
  while ((service = strsep (&p, " ")))
    {
      bool ok = true;
      if (level)
	ok = in_runlevel (level, service);
      else if (coldplugged)
	ok = (service_coldplugged (service)
	      && ! in_runlevel (softlevel, service)
	      && ! in_runlevel (bootlevel, service));

      if (! ok)
	continue;

      if (started)
	ok = (service_starting (service)
	      || service_started (service)
	      || service_stopping (service));
      else if (inactive)
	ok = service_inactive (service);

      if (! ok)
	continue;
    
      retval = true;
      lp = add_linkedlist (lp, service);
    }
  free (op);

  return retval;
}

/* Work out if a service is provided by another service.
   For example metalog provides logger.
   We need to be able to handle syslogd providing logger too.
   We do this by checking whats running, then what's starting/stopping,
   then what's run in the runlevels and finally alphabetical order.

   If there are any bugs in rc-depend, they will probably be here as
   provided dependancy can change depending on runlevel state.
   */
struct linkedlist *get_provided (struct depinfo *deptree,
				 struct depinfo *depinfo)
{
  if (exists_dir_file (INITDIR, depinfo->service))
    return NULL;

  struct deptype *dt = get_deptype (depinfo, "providedby");
  if (! dt)
    return NULL;

  struct linkedlist *providers = xmalloc (sizeof (struct linkedlist));
  struct linkedlist *lp = providers;
  memset (providers, 0, sizeof (struct linkedlist));
  char *p, *op, *service;

  /* If we are stopping then all depends are true, regardless of state.
     This is especially true for net services as they could force a restart
     of the local dns resolver which may depend on net. */
  if (is_runlevel_stop () || always_valid)
    {
      op = p = strdup (dt->services);
      while ((service = strsep (&p, " ")))
	lp = add_linkedlist (lp, service);
      free (op);

      if (providers->item)
	return (providers);

      free (providers);
      return NULL;
    }

  /* If we're strict, then only use what we have in our runlevel */
  if (strict)
    {
      op = p = strdup (dt->services);
      while ((service = strsep (&p, " ")))
	if (in_runlevel (softlevel, service))
	  lp = add_linkedlist (lp, service);
      free (op);

      if (providers->item)
	return (providers);
    }

  /* OK, we're not strict or there were no services in our runlevel.
     This is now where the logic gets a little fuzzy :)
     If there is >1 running service then we return NULL.
     We do this so we don't hang around waiting for inactive services and
     our need has already been satisfied as it's not strict.
     We apply this to our runlevel, coldplugged services, then bootlevel
     and finally any running.*/
#define DO \
  if (providers->next && providers->next->item) \
    { \
      free_linkedlist (providers); \
      return NULL; \
    } \
  else \
  return providers; \

  /* Anything in the runlevel has to come first */
  if (get_provided1 (providers, dt, softlevel, false, true, false))
    { DO }
  if (get_provided1 (providers, dt, softlevel, false, false, true))
    { DO }
  if (get_provided1 (providers, dt, softlevel, false, false, false))
    return providers;

  /* Check coldplugged started services */
  if (get_provided1 (providers, dt, NULL, true, true, false))
    { DO }

  /* Check bootlevel if we're not in it */
  if (strcmp (softlevel, bootlevel) != 0)
    {
      if (get_provided1 (providers, dt, bootlevel, false, true, false))
	{ DO }
      if (get_provided1 (providers, dt, bootlevel, false, false, true))
	{ DO }
    }

  /* Check coldplugged inactive services */
  if (get_provided1 (providers, dt, NULL, true, false, true))
    { DO }

  /* Check manually started */
  if (get_provided1 (providers, dt, NULL, false, true, false))
    { DO }
  if (get_provided1 (providers, dt, NULL, false, false, true))
    { DO }

  /* Nothing started then. OK, lets get the stopped services */
  if (get_provided1 (providers, dt, NULL, true, false, false))
    return providers;
  if ((strcmp (softlevel, bootlevel) != 0)
      && (get_provided1 (providers, dt, bootlevel, false, false, false)))
    return providers;

  /* Still nothing? OK, list all services */
  op = p = strdup (dt->services);
  while ((service = strsep (&p, " ")))
    lp = add_linkedlist (lp, service);
  free (op);

  if (providers->item)
    return (providers);

  /* Nothing provides us then */
  free_linkedlist (providers);
  return NULL;
}

void visit_service (struct depinfo *deptree,
		    struct linkedlist *types,
		    struct linkedlist *sorted, struct linkedlist *visited,
		    struct depinfo *depinfo,
		    bool descend)
{
  if (! deptree || !sorted || !visited || !depinfo)
    return;

  struct linkedlist *s;
  struct linkedlist *l = NULL;

  /* Check if we have already visited this service or not */
  for (s = visited; s; s = s->next)
    {
      if (s->item && strcmp (s->item, depinfo->service) == 0)
	return;
      l = s;
    }

  /* Add ourselves as a visited service */
  l = add_linkedlist(l, depinfo->service);

  char *p, *op;
  char *service;

  struct linkedlist *provides;
  struct linkedlist *type;
  struct linkedlist *lp;
  struct depinfo *di;
  struct deptype *dt;

  for (type = types; type; type = type->next)
    {
      if ((dt = get_deptype (depinfo, type->item)))
	{
	  op = p = strdup (dt->services);
	  while ((service = strsep (&p, " ")))
	    {
	      if (! descend || strcmp (type->item, "iprovide") == 0)
		{
		  add_linkedlist (sorted, service);
		  continue;
		}

	      di = get_depinfo (deptree, service);
	      if ((provides = get_provided (deptree, di)))
		{
		  for (lp = provides; lp; lp = lp->next)
		    {
		      di = get_depinfo (deptree, lp->item);
		      if (di && (strcmp (type->item, "ineed") == 0
				 || valid_service (di->service)))
			visit_service (deptree, types, sorted, visited, di,
				       true);
		    }
		  free_linkedlist (provides);
		}
	      else
		if (di && (strcmp (type->item, "ineed") == 0
			   || valid_service (service)))
		  visit_service (deptree, types, sorted, visited, di, true);
	    }
	  free (op);
	}
    }

  /* Now visit the stuff we provide for */
  if ((dt = get_deptype (depinfo, "iprovide")) && descend)
    {
      op = p = strdup (dt->services);
      while ((service = strsep (&p, " ")))
	{
	  if ((di = get_depinfo (deptree, service)))
	    if ((provides = get_provided (deptree, di)))
	      {
		for (lp = provides; lp; lp = lp->next)
		  if (strcmp (lp->item, depinfo->service) == 0)
		    {
		      visit_service (deptree, types, sorted, visited, di, true);
		      break;
		    }
		free_linkedlist (provides);
	      }
	}
      free (op);
    }

  /* We've visited everything we need, so add ourselves unless we
     are also the service calling us or we are provided by something */
  if (! svcname || strcmp (svcname, depinfo->service) != 0)
    if (! get_deptype (depinfo, "providedby"))
      add_linkedlist (sorted, depinfo->service);
}

int main (int argc, char **argv)
{

  int i;
  struct linkedlist *types = xmalloc (sizeof (struct linkedlist));
  struct linkedlist *services = xmalloc (sizeof (struct linkedlist));
  memset (types, 0, sizeof (struct linkedlist));
  memset (services, 0, sizeof (struct linkedlist));

  struct linkedlist *lasttype = types;
  struct linkedlist *lastservice = services;
  struct depinfo *di = NULL;
  bool trace = true;

  struct depinfo *deptree = NULL;
  for (i = 1; i < argc; i++)
    {
      if (strcmp (argv[i], "--alwaysvalid") == 0)
	{
	  always_valid = true;
	  continue;
	}

      if (strcmp (argv[i], "--strict") == 0)
	{
	  strict = true;
	  continue;
	}

      if (strcmp (argv[i], "--deptree") == 0)
	{
	  i++;
	  if (i == argc)
	    errx (EXIT_FAILURE, "no deptree specified");

	  if ((deptree = load_deptree (argv[i])) == NULL)
	    {
	      free_linkedlist (types);
	      free_linkedlist (services);
	      errx (EXIT_FAILURE, "failed to load deptree `%s'", argv[i]);
	    }

	  continue;
	}

      if (strcmp (argv[i], "--notrace") == 0)
	{
	  trace = false;
	  continue;
	}

      if (argv[i][0] == '-')
	{
	  argv[i]++;
	  lasttype = add_linkedlist (lasttype, argv[i]);
	}
      else
	{
	  if (! deptree && ((deptree = load_deptree (DEPTREE)) == NULL))
	    {
	      free_linkedlist (types);
	      free_linkedlist (services);
	      errx (EXIT_FAILURE, "failed to load deptree `%s'", DEPTREE);
	    }

	  di = get_depinfo (deptree, argv[i]);
	  if (! di)
	    warnx ("no dependency info for service `%s'", argv[i]);
	  else
	    lastservice = add_linkedlist (lastservice, argv[i]);
	}
    }

  if (! services->item)
    {
      free_linkedlist (types);
      free_linkedlist (services);
      free_deptree (deptree);
      errx (EXIT_FAILURE, "no services specified");
    }

  /* If we don't have any types, then supply some defaults */
  if (! types->item)
    {
      lasttype = add_linkedlist (lasttype, "ineed");
      lasttype = add_linkedlist (lasttype, "iuse");
    }

  /* Setup our runlevels */
  bootlevel = getenv ("BOOTLEVEL");
  if (! bootlevel)
    bootlevel = BOOTLEVEL;
  if (is_runlevel_stop ())
    softlevel = getenv ("OLDSOFTLEVEL");
  else
    softlevel = getenv ("SOFTLEVEL");
  if (! softlevel)
    softlevel = SOFTLEVEL;
  svcname = getenv ("SVCNAME");

  struct linkedlist *sorted = xmalloc (sizeof (struct linkedlist));
  struct linkedlist *visited = xmalloc (sizeof (struct linkedlist));
  memset (sorted, 0, sizeof (struct linkedlist));
  memset (visited, 0, sizeof (struct linkedlist));

  struct linkedlist *service;
  for (service = services; service; service = service->next)
    {
      di = get_depinfo (deptree, service->item);
      visit_service (deptree, types, sorted, visited, di, trace);
    }

  if (sorted->item)
    {
      struct linkedlist *s;
      bool first = true;
      for (s = sorted; s; s = s->next)
	{
	  if (first)
	    first = false;
	  else
	    printf (" ");
	  if (s->item)
	    printf ("%s", s->item);
	}
      printf ("\n");
    }

  free_linkedlist (sorted);
  free_linkedlist (visited);
  free_linkedlist (types);
  free_linkedlist (services);
  free_deptree (deptree);
  return (0);
}
